File: karabiner/src/rules.ts
import { mkdir, writeFile } from "node:fs/promises";
import type { KarabinerRules, Manipulator } from "./types.js";
import { createSubLayer, keyCode, app } from "./utils.js";
import { doubleCommandQ } from "./customRules/doubleCommandQ.js";
import { languageSwitch } from "./customRules/languageSwitch.js";

const hyperManipulator: Manipulator = {
  description: "Caps Lock -> Hyper Key",
  from: {
    key_code: "caps_lock",
    modifiers: {
      optional: ["any"],
    },
  },
  to: [
    {
      set_variable: {
        name: "hyper",
        value: 1,
      },
    },
    {
      key_code: "right_shift",
      modifiers: ["right_control", "left_option", "left_command"],
      lazy: true,
    },
  ],
  to_after_key_up: [
    {
      set_variable: {
        name: "hyper",
        value: 0,
      },
    },
  ],
  type: "basic",
};

type RulesOptions = {
  isLaptop: boolean;
};

const rules = ({ isLaptop }: RulesOptions) => [
  // Temporarily disabled — re-enable to restore Hyper.
  // {
  //   description: "Hyper Key (⌃⌥⇧⌘)",
  //   manipulators: [hyperManipulator],
  // },
  ...doubleCommandQ,
  ...languageSwitch({ isLaptop }),
  // ...navigationKeys,
  // ...deletionKeys,
  // https://github.com/pqrs-org/Karabiner-Elements/issues/2880#issuecomment-1774847928
  {
    description: "Temporary Fix for sleep issue",
    manipulators: [
      {
        type: "basic",
        from: {
          key_code: "escape",
        },
        to_if_alone: [
          {
            key_code: "escape",
          },
        ],
      },
    ],
  },
  createSubLayer("right_option", "Media Commands Sublayer + Apps", {
    s: keyCode("play_or_pause"),
    d: keyCode("fastforward"),
    a: keyCode("rewind"),
    t: app("WezTerm"),
    g: app("GoLand"),
    w: app("WebStorm"),
    b: app("Dia"),
    z: app("Zed"),
    l: app("Logseq"),
    m: app("Telegram"),
    h: app("Bruno v3 Preview"),
    n: app("Obsidian"),
    f: app("Finder"),
    c: app("Claude"),
    4: app("Sublime Merge"),
  }),
];

const isLaptop = process.argv.includes("--laptop");

const fileContent = JSON.stringify(
  {
    global: {
      show_in_menu_bar: false,
    },
    profiles: [
      {
        name: "Default",
        selected: true,
        complex_modifications: {
          rules: rules({ isLaptop }),
        },
      },
    ],
  },
  null,
  2,
);

const outputFolder = "./dist";
const outputFile = `${outputFolder}/karabiner.json`;

await mkdir(outputFolder, { recursive: true });
await writeFile(outputFile, fileContent, "utf8");

console.log(`Generated Karabiner rules in ${outputFile}`);

File: karabiner/rules.go
package main

type layerCmd struct {
	to []to
}

type layerEntry struct {
	key string
	cmd layerCmd
}

func keyCode(code string) layerCmd {
	return layerCmd{to: []to{{KeyCode: code}}}
}

func subLayer(modifier, description string, entries []layerEntry) rule {
	manipulators := make([]manipulator, len(entries))
	for i, e := range entries {
		manipulators[i] = manipulator{
			Type: "basic",
			From: from{
				KeyCode:   e.key,
				Modifiers: &modifiers{Mandatory: []string{modifier}},
			},
			To: e.cmd.to,
		}
	}
	return rule{Description: description, Manipulators: manipulators}
}

func intPtr(v int) *int { return &v }

func doubleCommandQ() rule {
	commandQ := from{KeyCode: "q", Modifiers: &modifiers{Mandatory: []string{"command"}}}
	reset := []to{{SetVariable: &setVariable{Name: "command-q", Value: 0}}}
	return rule{
		Description: "Double Command+Q to quit",
		Manipulators: []manipulator{
			{
				Type:       "basic",
				From:       commandQ,
				To:         []to{{KeyCode: "q", Modifiers: []string{"left_command"}}},
				Conditions: []condition{{Type: "variable_if", Name: "command-q", Value: intPtr(1)}},
			},
			{
				Type: "basic",
				From: commandQ,
				To:   []to{{SetVariable: &setVariable{Name: "command-q", Value: 1}}},
				ToDelayedAction: &delayedAction{
					ToIfInvoked:  reset,
					ToIfCanceled: reset,
				},
			},
		},
	}
}

// moji owns the layout: Karabiner only decides tap versus hold per device and emits f19 on a
// tap, which moji swallows, switching the layout and holding the keys typed behind the switch.
const signalKey = "f19"

const (
	corneVendorID  = 18003
	corneProductID = 4
)

// shift is held constantly while typing, so only a deliberate short tap may switch the layout
const shiftTapTimeout = 180

type langVariant struct {
	from   from
	to     []to
	device *condition
	params *parameters
}

func (v langVariant) manipulator() manipulator {
	var conditions []condition
	if v.device != nil {
		conditions = append(conditions, *v.device)
	}
	return manipulator{
		Type:       "basic",
		From:       v.from,
		Conditions: conditions,
		ToIfAlone:  []to{{KeyCode: signalKey}},
		To:         v.to,
		Parameters: v.params,
	}
}

func languageSwitch() rule {
	noRepeat := false
	globe := manipulator{
		Type: "basic",
		From: from{AppleVendorTopCaseKeyCode: "keyboard_fn"},
		To:   []to{{KeyCode: signalKey, Repeat: &noRepeat}},
	}
	variants := []langVariant{
		{
			from:   from{KeyCode: "left_control"},
			to:     []to{{KeyCode: "left_control"}},
			device: &condition{Type: "device_unless", Identifiers: []identifiers{{IsBuiltInKeyboard: true}}},
		},
		{
			from:   from{KeyCode: "left_shift"},
			to:     []to{{KeyCode: "left_shift"}},
			device: &condition{Type: "device_if", Identifiers: []identifiers{{VendorID: corneVendorID, ProductID: corneProductID}}},
			params: &parameters{ToIfAloneTimeout: shiftTapTimeout},
		},
	}
	manipulators := []manipulator{globe}
	for _, v := range variants {
		manipulators = append(manipulators, v.manipulator())
	}
	return rule{Description: "Switch the keyboard layout through moji", Manipulators: manipulators}
}

func hyperKey() rule {
	return rule{
		Description: "Caps Lock is Hyper",
		Manipulators: []manipulator{
			{
				Type: "basic",
				From: from{KeyCode: "caps_lock", Modifiers: &modifiers{Optional: []string{"any"}}},
				To:   []to{{KeyCode: "left_shift", Modifiers: []string{"left_command", "left_option", "left_control"}}},
			},
		},
	}
}

func f5ToF13() rule {
	return rule{
		Description: "F5 -> F13",
		Manipulators: []manipulator{
			{
				Type: "basic",
				From: from{KeyCode: "f5", Modifiers: &modifiers{Optional: []string{"any"}}},
				To:   []to{{KeyCode: "f13"}},
			},
		},
	}
}

func f6ToF18() rule {
	return rule{
		Description: "F6 -> F18",
		Manipulators: []manipulator{
			{
				Type: "basic",
				From: from{KeyCode: "f6", Modifiers: &modifiers{Optional: []string{"any"}}},
				To:   []to{{KeyCode: "f18"}},
			},
		},
	}
}

func escapeSleepFix() rule {
	return rule{
		Description: "Temporary Fix for sleep issue",
		Manipulators: []manipulator{
			{
				Type:      "basic",
				From:      from{KeyCode: "escape"},
				ToIfAlone: []to{{KeyCode: "escape"}},
			},
		},
	}
}

func mediaSubLayer() rule {
	return subLayer("right_option", "Media Commands Sublayer", []layerEntry{
		{key: "s", cmd: keyCode("play_or_pause")},
		{key: "d", cmd: keyCode("fastforward")},
		{key: "a", cmd: keyCode("rewind")},
	})
}
